File: Java/jobspls/algorithms/MaxCounters.java
package jobspls.algorithms;

public class MaxCounters {
    public static int[] max_counters(int n_counters, int[] operations) {
        var counters = new int[n_counters];
        int cur_max = 0;
        int last_max = 0;

        for (var i = 0; i < operations.length; i++) {
            var op = operations[i];
            if (op < n_counters + 1) {
                var current = counters[op - 1];
                if (current < last_max) {
                    current = last_max;
                }
                current++;

                if (current > cur_max) {
                    cur_max = current;
                }
                
                counters[op - 1] = current;
            } else {
                last_max = cur_max;
            }
        }

        for (var i = 0; i < n_counters; i++) {
            if (counters[i] < last_max) {
                counters[i] = last_max;
            }
        }

        return counters;
    }
}

File: Go/jobspls/algorithms/max_counters.go
package algorithms

func MaxCounters(n_counters int, operations []int) []int {
	counters := make([]int, n_counters)
	cur_max := 0
	last_max := 0

	for _, op := range operations {
		if op < n_counters+1 {
			// Increase counter operation
			current := counters[op-1]

			// If current is smaller than last_max, the value at this index
			// hasn't been updated to last_max yet
			if current < last_max {
				current = last_max
			}
			current += 1

			counters[op-1] = current
			if current > cur_max {
				cur_max = current
			}
		} else {
			last_max = cur_max
		}
	}

	// Update any counter values that should be set to the maximum
	for i, counter := range counters {
		if counter < last_max {
			counters[i] = last_max
		}
	}

	return counters
}
